flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala

```diff
@@ -18,231 +18,574 @@
+
+    // case class
+    util.verifySchema(
+      util.addTable[CClassWithTime]('cf1, 'cf3.proctime, 'cf2),
+      Seq("cf1" -> INT, "cf3" -> PROCTIME, "cf2" -> LONG))
+
+    util.verifySchema(
+      util.addTable[CClassWithTime]('cf1, 'cf3.rowtime, 'cf2),
+      Seq("cf1" -> INT, "cf3" -> ROWTIME, "cf2" -> LONG))
+
+    // row
+    util.verifySchema(
+      util.addTable('rf1, 'rf3.proctime, 'rf2)(TEST_ROW_WITH_TIME),
+      Seq("rf1" -> INT, "rf3" -> PROCTIME, "rf2" -> LONG))
+
+    util.verifySchema(
+      util.addTable('rf1, 'rf3.rowtime, 'rf2)(TEST_ROW_WITH_TIME),
+      Seq("rf1" -> INT, "rf3" -> ROWTIME, "rf2" -> LONG))
+
+    // tuple
+    util.verifySchema(
+      util.addTable[JTuple3[Int, Long, String]]('f0, 'f2.proctime, 'f1),
+      Seq("f0" -> INT, "f2" -> PROCTIME, "f1" -> LONG))
+
+    util.verifySchema(
+      util.addTable[JTuple3[Int, Long, String]]('f0, 'f2.rowtime, 'f1),
+      Seq("f0" -> INT, "f2" -> ROWTIME, "f1" -> LONG))
   }
 
   @Test
-  def testGetFieldInfoTupleNames(): Unit = {
-    val fieldInfo = tEnv.getFieldInfo(
-      tupleType,
-      Array(
-        UnresolvedFieldReference("name1"),
-        UnresolvedFieldReference("name2"),
-        UnresolvedFieldReference("name3")
-      ))
-
-    fieldInfo._1.zip(Array("name1", "name2", "name3")).foreach(x => assertEquals(x._2, x._1))
-    fieldInfo._2.zip(Array(0, 1, 2)).foreach(x => assertEquals(x._2, x._1))
+  def testAliasByName(): Unit = {
+    val utils = Seq(streamTestUtil(), batchTestUtil())
+
+    utils.foreach { util =>
+
+      // atomic
+      util.verifySchema(
+        util.addTable[Int]('myint as 'new),
+        Seq("new" -> INT))
+
+      // case class
+      util.verifySchema(
+        util.addTable[CClass]('cf1, 'cf3 as 'new, 'cf2),
+        Seq("cf1" -> INT, "new" -> DOUBLE, "cf2" -> STRING))
+
+      // row
+      util.verifySchema(
+        util.addTable('rf1, 'rf3 as 'new, 'rf2)(TEST_ROW),
+        Seq("rf1" -> INT, "new" -> DOUBLE, "rf2" -> STRING))
+
+      // tuple
+      util.verifySchema(
+        util.addTable[JTuple3[Int, String, Double]]('f0, 'f2 as 'new, 'f1),
+        Seq("f0" -> INT, "new" -> DOUBLE, "f1" -> STRING))
+    }
   }
 
   @Test
-  def testGetFieldInfoCClassNames(): Unit = {
-    val fieldInfo = tEnv.getFieldInfo(
-      caseClassType,
-      Array(
-        UnresolvedFieldReference("name1"),
-        UnresolvedFieldReference("name2"),
-        UnresolvedFieldReference("name3")
-      ))
-
-    fieldInfo._1.zip(Array("name1", "name2", "name3")).foreach(x => assertEquals(x._2, x._1))
-    fieldInfo._2.zip(Array(0, 1, 2)).foreach(x => assertEquals(x._2, x._1))
+  def testStreamAliasWithAddingTimeAttributesByName(): Unit = {
+    val util = streamTestUtil()
+
+    // atomic
+    util.verifySchema(
+      util.addTable[Int](('myint as 'new).proctime),
+      Seq("new" -> PROCTIME))
+
+    // case class
+    util.verifySchema(
+      util.addTable[CClassWithTime]('cf1, ('newnew as 'new).proctime, 'cf2),
```

Comment:
I think we should add checks that `'newnew` is actually present in the schema. For `proctime` `as` should be forbidden, IMO.